src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp

```diff
@@ -513,71 +513,64 @@ class ShenandoahCardCluster: public CHeapObj<mtGC> {
 
-  uint16_t *object_starts;
+  // ObjectStartsInCardRegion bit is set within a crossing_info.offsets.start iff at least one object starts within
+  // a particular card region.  We pack this bit into start byte under assumption that start byte is accessed less
+  // frequently that last byte.  This is true when number of clean cards is greater than number of dirty cards.
+  static const uint16_t ObjectStartsInCardRegion = 0x80;
+  static const uint16_t FirstStartBits           = 0x3f;
+
+  crossing_info *object_starts;
 
 public:
+  // If we're setting first_start, assume the card has an object.
   inline void set_first_start(size_t card_index, uint8_t value) {
-    object_starts[card_index] &= ~FirstStartBits;
-    object_starts[card_index] |= (FirstStartBits & (value << FirstStartShift));
+    object_starts[card_index].offsets.first = ObjectStartsInCardRegion | value;
   }
 
   inline void set_last_start(size_t card_index, uint8_t value) {
-    object_starts[card_index] &= ~LastStartBits;
-    object_starts[card_index] |= (LastStartBits & (value << LastStartShift));
+    object_starts[card_index].offsets.last = value;
   }
 
   inline void set_has_object_bit(size_t card_index) {
-    object_starts[card_index] |= ObjectStartsInCardRegion;
+    object_starts[card_index].offsets.first |= ObjectStartsInCardRegion;
   }
 
   inline void clear_has_object_bit(size_t card_index) {
-    object_starts[card_index] &= ~ObjectStartsInCardRegion;
+    object_starts[card_index].offsets.first &= ~ObjectStartsInCardRegion;
+  }
+
+  // Returns true iff an object is known to start within the card memory associated with card card_index.
+  inline bool has_object(size_t card_index) {
+    return (object_starts[card_index].offsets.first & ObjectStartsInCardRegion) != 0;
   }
 
   inline void clear_objects_in_range(HeapWord *addr, size_t num_words) {
     size_t card_index = _rs->card_index_for_addr(addr);
     size_t last_card_index = _rs->card_index_for_addr(addr + num_words - 1);
     while (card_index <= last_card_index)
-      object_starts[card_index++] = 0;
+      object_starts[card_index++].short_word = 0;
   }
 
   ShenandoahCardCluster(RememberedSet *rs) {
     _rs = rs;
     // TODO: We don't really need object_starts entries for every card entry.  We only need these for
     // the card entries that correspond to old-gen memory.  But for now, let's be quick and dirty.
-    object_starts = (uint16_t *) malloc(rs->total_cards() * sizeof(uint16_t));
-    if (object_starts == NULL)
+    object_starts = (crossing_info *) malloc(rs->total_cards() * sizeof(crossing_info));
+    if (object_starts == nullptr)
       fatal("Insufficient memory for initializing heap");
     for (size_t i = 0; i < rs->total_cards(); i++)
-      object_starts[i] = 0;
+      object_starts[i].short_word = 0;
   }
 
   ~ShenandoahCardCluster() {
-    if (object_starts != NULL)
+    if (object_starts != nullptr)
       free(object_starts);
```

Comment:
I know `ShenandoahCardCluster` is a singleton so this is a nit pick, but we should set `object_starts` to `nullptr` here to avoid double deletes.

Reply:
fixed.